## **Word2Vec**

1.  Word2Vec is a word embedding technique that is used to convert each and every input word into the dense dimensional numerical vector so that it can easily be trained by the machine learning model while preserving its semantic meaning.

2.  Words with the similar meaning will have the same numerical vector valued representations.

## **Problem Stetement:**

1.  The main aim of the problem statement is to find out the vector representation of each and every word and find out the similar words w.r.to the particular word and even compute the similarity ratio between the two words.


## **Step 1: Import all the necessary libraries**

In [194]:
!pip install gensim

In [195]:
import  numpy               as  np
import  pandas              as  pd
import  matplotlib.pyplot   as  plt
import  seaborn             as  sns

import  nltk

from    nltk.tokenize       import  word_tokenize
from    gensim.models       import  Word2Vec

#### OBSERVATIONS:

1.  gensim --------------->   A model used for embedding

2.  word_tokenize -------->  converts the sentences into the words

## **Step 2: Download the necessary resources required for NLTK**

In [196]:
#### Download the necessary resources
nltk.download('punkt_tab')
nltk.download('averaged_perceptron_tagger_eng')
nltk.download('stopwords')

[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger_eng to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger_eng is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

## **OBSERVATIONS:**

1.  punkt_tab ---------->  It is a pre trained tokenizer required for tokenization process.

2.  averaged_perceptron_tagger_eng --------> It is needed for POS Tagging.

3.  stopwords -----------> It is needed for the removal of the stop words

## **Step 3: Load the Dataset**

In [197]:
### Load the dataset

df = pd.read_csv('/content/SMSSpamCollection',sep='\t',names=['label','message'])

In [198]:
df

,label,message
0,ham,"Go until jurong point, crazy.. Available only ..."
1,ham,Ok lar... Joking wif u oni...
2,spam,Free entry in 2 a wkly comp to win FA Cup fina...
3,ham,U dun say so early hor... U c already then say...
4,ham,"Nah I don't think he goes to usf, he lives aro..."
...,...,...
5567,spam,This is the 2nd time we have tried 2 contact u...
5568,ham,Will ü b going to esplanade fr home?
5569,ham,"Pity, * was in mood for that. So...any other s..."
5570,ham,The guy did some bitching but I acted like i'd...


In [199]:
#### Interchange the positions of the message and the labels

df=df[['message','label']]

In [200]:
df

,message,label
0,"Go until jurong point, crazy.. Available only ...",ham
1,Ok lar... Joking wif u oni...,ham
2,Free entry in 2 a wkly comp to win FA Cup fina...,spam
3,U dun say so early hor... U c already then say...,ham
4,"Nah I don't think he goes to usf, he lives aro...",ham
...,...,...
5567,This is the 2nd time we have tried 2 contact u...,spam
5568,Will ü b going to esplanade fr home?,ham
5569,"Pity, * was in mood for that. So...any other s...",ham
5570,The guy did some bitching but I acted like i'd...,ham


## **OBSERVATIONS:**

1.  The dataset has been created with two columns as follows:---

    (a.)  message -------->  It specifies the type of the email message.

    (b.)  label   -------->  It specifies the type of the message whether it is 'ham' or 'spam'.

## **Step 4: Perform Text Preprocessing**

In [201]:
### perform word tokenization on all the sentences in the message column

res = [word_tokenize(x.lower()) for x in df['message']]

In [202]:
res

[['go',
  'until',
  'jurong',
  'point',
  ',',
  'crazy',
  '..',
  'available',
  'only',
  'in',
  'bugis',
  'n',
  'great',
  'world',
  'la',
  'e',
  'buffet',
  '...',
  'cine',
  'there',
  'got',
  'amore',
  'wat',
  '...'],
 ['ok', 'lar', '...', 'joking', 'wif', 'u', 'oni', '...'],
 ['free',
  'entry',
  'in',
  '2',
  'a',
  'wkly',
  'comp',
  'to',
  'win',
  'fa',
  'cup',
  'final',
  'tkts',
  '21st',
  'may',
  '2005.',
  'text',
  'fa',
  'to',
  '87121',
  'to',
  'receive',
  'entry',
  'question',
  '(',
  'std',
  'txt',
  'rate',
  ')',
  't',
  '&',
  'c',
  "'s",
  'apply',
  '08452810075over18',
  "'s"],
 ['u',
  'dun',
  'say',
  'so',
  'early',
  'hor',
  '...',
  'u',
  'c',
  'already',
  'then',
  'say',
  '...'],
 ['nah',
  'i',
  'do',
  "n't",
  'think',
  'he',
  'goes',
  'to',
  'usf',
  ',',
  'he',
  'lives',
  'around',
  'here',
  'though'],
 ['freemsg',
  'hey',
  'there',
  'darling',
  'it',
  "'s",
  'been',
  '3',
  'week',
  "'s",
  'n

## **OBSERVATIONS**

1.  The above list is a list of all the tokenized sentences obtained from the message column of the dataframe.

2.  This can be fed as an input into the Word2Vec model.

## **Step 5: Train the Word2Vec model**

In [203]:
### Train the Word2Vec model

model = Word2Vec(
    sentences   =    res            ,
    vector_size =     50            ,
    window      =     5             ,
    min_count   =     1             ,
    sg          =     0             ,
    workers     =     1
)

## **OBSERVATIONS:**

1.  The model of Word2Vec has been created with the help of the following parameters:-----

    (a.)  sentences = list of the tokenized sentences

    (b.)  vector_size = size of the dense dimensional embedded vector

    (c.)  window = 5 ---->  5 context input neighbors need to be considered for determining the target

    (d.)  min_count = 1 ------>  Minimum number of frequency needed for the dictionary

    (e.)sg = 0 ------------> determines the type of the model architecture.

            (i.)  If sg = 0 ----------->  Model architecture is CBOW.


2.The CBOW model is trained.

In [204]:
model

## **Step 6: Get the vocabulary from the model**

In [205]:
#### Get the vocabulary

vocabulary = model.wv.key_to_index

In [206]:
vocabulary

{'.': 0,
 'i': 1,
 'to': 2,
 'you': 3,
 ',': 4,
 '?': 5,
 'a': 6,
 '!': 7,
 'the': 8,
 '...': 9,
 'u': 10,
 'and': 11,
 '&': 12,
 'is': 13,
 'in': 14,
 'me': 15,
 ';': 16,
 'my': 17,
 'it': 18,
 ':': 19,
 'for': 20,
 '..': 21,
 'your': 22,
 'of': 23,
 'have': 24,
 'call': 25,
 'that': 26,
 'do': 27,
 'on': 28,
 ')': 29,
 'are': 30,
 "'s": 31,
 'now': 32,
 '2': 33,
 'so': 34,
 'not': 35,
 'but': 36,
 'or': 37,
 'can': 38,
 'at': 39,
 'we': 40,
 "'m": 41,
 'ur': 42,
 'if': 43,
 'get': 44,
 'will': 45,
 'be': 46,
 'with': 47,
 'just': 48,
 "n't": 49,
 'no': 50,
 'this': 51,
 'gt': 52,
 'lt': 53,
 '*': 54,
 'up': 55,
 'how': 56,
 'when': 57,
 'what': 58,
 '4': 59,
 '#': 60,
 'from': 61,
 'go': 62,
 'free': 63,
 'all': 64,
 'out': 65,
 'know': 66,
 'got': 67,
 'ok': 68,
 'like': 69,
 'good': 70,
 'was': 71,
 'then': 72,
 '-': 73,
 'come': 74,
 "'ll": 75,
 'there': 76,
 'he': 77,
 'am': 78,
 'its': 79,
 'time': 80,
 'only': 81,
 'day': 82,
 'love': 83,
 'want': 84,
 'send': 85,
 'text': 86,


## **OBSERVATIONS**

1.  Here we have obtained the list of all the vocabularies after getting trained from Word2Vec model.

## **Step 7: Get the word vector**

In [207]:
#### Get the vector form of the word 'easy'

model.wv['easy']

array([ 0.09734271, -0.01055117, -0.01710207, -0.03041005, -0.04830376,
       -0.1648974 ,  0.22502361,  0.3761735 , -0.21187302, -0.13272245,
        0.04628559, -0.31011316,  0.15573879,  0.12620212, -0.17804702,
        0.22563155,  0.20656124,  0.04704044, -0.3272515 , -0.24987622,
        0.12505724,  0.06921759,  0.50484574, -0.110344  ,  0.18428017,
        0.10418702, -0.06650363,  0.04961716, -0.27826208,  0.15405284,
        0.00287578, -0.0374819 ,  0.03847683, -0.04180386, -0.01649134,
        0.06086094,  0.2371307 , -0.02972472,  0.05957698, -0.08381695,
        0.24758384, -0.01027965, -0.02863646,  0.01659336,  0.4151543 ,
        0.07292053, -0.02229223, -0.20389405,  0.21386823,  0.12522535],
      dtype=float32)

In [208]:
#### Get the vector form of the word 'free'
model.wv['free']

array([ 0.50413895,  0.05331755,  0.18743967,  0.3283631 , -0.2111056 ,
       -0.39361933,  1.2648715 ,  1.5923188 , -1.1976776 , -0.4676762 ,
        0.3056324 , -1.5570214 ,  0.9978619 ,  0.30205086, -1.0110929 ,
        0.6423959 ,  1.1471316 ,  0.02725804, -1.4921259 , -1.4002966 ,
        0.9363259 ,  0.32259637,  2.6252425 , -0.6402094 ,  0.5346676 ,
        0.02925752, -0.41576242,  0.05053121, -1.2300285 ,  0.58581376,
        0.32910737, -0.01759705,  0.56483066,  0.03677833, -0.03192168,
        0.07478218,  0.60970646,  0.08529624,  0.04305375, -0.5105338 ,
        1.3178703 , -0.26613316, -0.17343721, -0.16297697,  1.4071605 ,
       -0.11222028,  0.330627  , -0.82567924,  0.9197642 ,  0.43713483],
      dtype=float32)

## **OBSERVATIONS**

1.  Here we have obtained the dense dimensional numerical vector of the size 50 for the word 'easy' and 'free'.

## **Step 8: Get the Vector Shape**

In [209]:
#### Get the shape of the vector

model.wv['easy'].shape

(50,)

In [210]:
model.wv['gets'].shape

(50,)

## **OBSERVATIONS:**

1.  The shape of the input vector is (50,).

## **Step 9: Get the most similar words w.r.to a particular word**

In [211]:
#### Get the most similar word w.r.to 'easy'
model.wv.most_similar('easy')

[('phone', 0.9965216517448425),
 ('more', 0.9959041476249695),
 ('between', 0.9958078265190125),
 ('chat', 0.9957349896430969),
 ('number', 0.9956907033920288),
 ('waiting', 0.99561607837677),
 ('thanks', 0.9955706000328064),
 ('fucking', 0.9954506158828735),
 ('off', 0.9954097867012024),
 ('with', 0.9953771829605103)]

## **OBSERVATIONS:**

1.  The above are the list of all the words that are similar to easy as their vector values are verymuch close to the vector representation value of 'easy'.

In [212]:
#### Get the most similar word w.r.to 'free'
model.wv.most_similar('free')

[('nokia', 0.9978139400482178),
 ('mobile', 0.9969006180763245),
 ('claim', 0.9961849451065063),
 ('txt', 0.995154619216919),
 ('latest', 0.9947271347045898),
 ('mobileupd8', 0.9938870072364807),
 ('camcorder', 0.993559718132019),
 ('orange', 0.9935161471366882),
 ('cash', 0.9933136105537415),
 ('entry', 0.993053674697876)]

## **OBSERVATIONS:**

1.  The above are the list of all the words that are similar to easy as their vector values are verymuch close to the vector representation value of 'free'.

## **Step 11: Get the Similarity between two words:**

In [213]:
#### Get the similarity between two words
model.wv.similarity('free','win')

np.float32(0.9730826)

## **OBSERVATIONS:**

1.  Here the similarity score between the two words ('free','win') is 0.97 which states that these two words are more identical with respect to each other.